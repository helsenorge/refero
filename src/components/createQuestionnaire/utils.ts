import Group from '@components/formcomponents/group/group';
import Display from '@formcomponents/display/display';
import Text from '@formcomponents/text/text';
import { QuestionnaireResponseItem, QuestionnaireItem } from 'fhir/r4';

import Attachment from '@formcomponents/attachment/attachment';
import Boolean from '@formcomponents/boolean/boolean';
import Choice from '@formcomponents/choice/choice';
import Date from '@formcomponents/date/date';
import DateTimeInput from '@formcomponents/date/date-time';
import Time from '@formcomponents/date/time';
import Decimal from '@formcomponents/decimal/decimal';
import Integer from '@formcomponents/integer/integer';
import OpenChoice from '@formcomponents/open-choice/open-choice';
import Quantity from '@formcomponents/quantity/quantity';
import StringComponent from '@formcomponents/string/string';
import TableContainer from '@formcomponents/table/TableContainer';

import { ComponentType } from 'react';
import { QuestionnaireComponentItemProps } from '@/components/createQuestionnaire/GenerateQuestionnaireComponents';
import { isRepeat, isTableCode } from '@/util';
import { createIdSuffix, getItemWithIdFromResponseItemArray, getRootQuestionnaireResponseItemFromData, Path } from '@/util/refero-core';
import { FormData } from '@/reducers/form';
import ItemType from '@/constants/itemType';

export const getResponseItems = (
  item?: QuestionnaireItem,
  formData?: FormData | null,
  path?: Path[]
): QuestionnaireResponseItem[] | undefined => {
  if (!formData || !formData.Content || !formData.Content.item) {
    return undefined;
  }
  const responseItemsFromData = getRootQuestionnaireResponseItemFromData(item?.linkId, formData);

  if (responseItemsFromData) {
    return responseItemsFromData;
  }
  // Navigate to the response item corresponding to the current path
  const responseItem = getResponseItemWithPath(path, formData.Content.item);

  if (!responseItem) {
    return undefined;
  }

  if (responseItem.linkId === item?.linkId) {
    // If the response item matches the current item, return it
    return [responseItem];
  } else {
    // Otherwise, search within its child items
    const childItems = responseItem.item;
    const childAnswers = responseItem.answer?.flatMap(ans => ans.item || []);

    const matchingItems = getItemWithIdFromResponseItemArray(item?.linkId, [...(childItems || []), ...(childAnswers || [])]);

    return matchingItems;
  }
};
function getResponseItemWithPath(path: Path[] = [], items: QuestionnaireResponseItem[]): QuestionnaireResponseItem | undefined {
  let currentItems = items;
  let responseItem: QuestionnaireResponseItem | undefined;

  for (const pathPart of path) {
    const index = pathPart.index ?? 0;

    const matchingItems = currentItems.filter(item => item.linkId === pathPart.linkId);
    if (matchingItems.length <= index) {
      return undefined;
    }

    responseItem = matchingItems[index];

    if (!responseItem) {
      return undefined;
    }

    // Prepare for the next iteration
    currentItems = responseItem.item || responseItem.answer?.flatMap(ans => ans.item || []) || [];
  }

  return responseItem;
}
export function getComponentForItem(
  type: string,
  extensionCode?: string | string[]
): ComponentType<QuestionnaireComponentItemProps> | undefined {
  if (String(type) === ItemType.GROUP && !!extensionCode && isTableCode(extensionCode)) {
    return TableContainer;
  } else if (String(type) === ItemType.GROUP) {
    return Group;
  }
  if (String(type) === ItemType.DISPLAY) {
    return Display;
  }
  if (String(type) === ItemType.BOOLEAN) {
    return Boolean;
  }
  if (String(type) === ItemType.DECIMAL) {
    return Decimal;
  }
  if (String(type) === ItemType.INTEGER) {
    return Integer;
  }
  if (String(type) === ItemType.DATE) {
    return Date;
  }
  if (String(type) === ItemType.DATETIME) {
    return DateTimeInput;
  }
  if (String(type) === ItemType.TIME) {
    return Time;
  }
  if (String(type) === ItemType.STRING) {
    return StringComponent;
  }
  if (String(type) === ItemType.TEXT) {
    return Text;
  }
  if (String(type) === ItemType.CHOICE) {
    return Choice;
  }
  if (String(type) === ItemType.OPENCHOICE) {
    return OpenChoice;
  }
  if (String(type) === ItemType.ATTATCHMENT) {
    return Attachment;
  }
  if (String(type) === ItemType.QUANTITY) {
    return Quantity;
  }
  return undefined;
}

export const createIdFormComponentIds = (item?: QuestionnaireItem, path?: Path[], index?: number): string =>
  `${item?.linkId}-${createIdSuffix(path, index, isRepeat(item))}`;
