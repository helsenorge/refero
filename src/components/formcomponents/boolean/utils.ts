import { QuestionnaireItem } from 'fhir/r4';
import { UseFormRegisterReturn } from 'react-hook-form';

import { isRequired } from '../../../util';
import { getValidationTextExtension } from '../../../util/extension';
import { Resources } from '../../../util/resources';
import { FormProps } from '../../../validation/ReactHookFormHoc';

export const booleanFormRegister = (
  register: FormProps['register'],
  item: QuestionnaireItem,
  resources?: Resources,
  value?: boolean,
  onChange?: () => void
): UseFormRegisterReturn => {
  return register(item.linkId, {
    required: {
      value: isRequired(item),
      message: resources?.formRequiredErrorMessage ?? 'Feltet er p√•krevd',
    },
    value,
    onChange,
  });
};
