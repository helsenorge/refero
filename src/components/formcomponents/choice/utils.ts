import { QuestionnaireItem } from 'fhir/r4';
import { UseFormRegisterReturn } from 'react-hook-form';

import { isRequired } from '../../../util';
import { getValidationTextExtension } from '../../../util/extension';
import { Resources } from '../../../util/resources';
import { FormProps } from '../../../validation/ReactHookFormHoc';

export const selectFormRegister = (
  register: FormProps['register'],
  item: QuestionnaireItem,
  resources?: Resources,
  value?: string,
  onChange?: (e: React.ChangeEvent<HTMLSelectElement>) => void,
  disabled?: boolean
): UseFormRegisterReturn => {
  return register(item.linkId, {
    value,
    onChange,
    required: {
      message: getValidationTextExtension(item) ?? resources?.formRequiredErrorMessage ?? '',
      value: isRequired(item),
    },
    disabled,
  });
};
