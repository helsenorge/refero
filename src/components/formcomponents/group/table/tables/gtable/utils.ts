import { format, isValid } from 'date-fns';
import {
  Coding,
  Extension,
  Quantity,
  Questionnaire,
  QuestionnaireItem,
  QuestionnaireResponse,
  QuestionnaireResponseItem,
  QuestionnaireResponseItemAnswer,
} from 'fhir/r4';

import { DATEFORMATS } from '../constants';

import { TIME_SEPARATOR } from '@/constants/dateTimeConstants';
import { COPY_EXPRESSION_URL } from '@/constants/extensions';
import { getResponseItemsWithLinkId } from '@/util/refero-core';

/* -------------------------------------------------------------------------- */
/* TYPES                                                                      */
/* -------------------------------------------------------------------------- */

export type Column = {
  header: string;
  sourceLinkId: string;
  summaryLinkId: string;
};

type QuantityKeys = keyof Pick<Quantity, 'value' | 'code' | 'system' | 'unit'> | 'display';
type CodingKeys = keyof Pick<Coding, 'code' | 'display' | 'system'>;

/* -------------------------------------------------------------------------- */
/* FHIRPATH / EXPRESSION HELPERS                                              */
/* -------------------------------------------------------------------------- */

export function extractLinkIdFromExpression(expr: string): string | undefined {
  const match = expr.match(/linkId\s*=\s*['"]([^'"]+)['"]/);
  return match?.[1];
}

const getCopyTargetFhirPath = (extensions?: Extension[]): string | undefined =>
  extensions?.find(ext => ext.url === COPY_EXPRESSION_URL)?.valueString;

/**
 * Extract columns from a gTable group item
 */
export function getColumnsFromGTableItem(gTableItem: QuestionnaireItem): Column[] {
  const cols: Column[] = [];
  for (const cell of gTableItem.item ?? []) {
    const expr = getCopyTargetFhirPath(cell.extension);
    if (!expr) continue;
    const source = extractLinkIdFromExpression(expr);
    if (!source) continue;

    cols.push({
      header: cell.text ?? cell.linkId ?? '(kolonne)',
      sourceLinkId: source,
      summaryLinkId: cell.linkId ?? '',
    });
  }
  return cols;
}

/* -------------------------------------------------------------------------- */
/* QUESTIONNAIRE STRUCTURE                                                    */
/* -------------------------------------------------------------------------- */

export function walkQ(
  items: QuestionnaireItem[] | undefined,
  fn: (item: QuestionnaireItem, parents: QuestionnaireItem[]) => void,
  parents: QuestionnaireItem[] = []
): void {
  if (!items) return;
  for (const item of items) {
    fn(item, parents);
    if (item.item?.length) walkQ(item.item, fn, [...parents, item]);
  }
}

export function indexQuestionnaire(q: Questionnaire | undefined | null): {
  byLinkId: Map<string, QuestionnaireItem>;
  parentPath: Map<string, QuestionnaireItem[]>;
} {
  const byLinkId = new Map<string, QuestionnaireItem>();
  const parentPath = new Map<string, QuestionnaireItem[]>();
  walkQ(q?.item, (item, parents) => {
    if (item.linkId) {
      byLinkId.set(item.linkId, item);
      parentPath.set(item.linkId, parents);
    }
  });

  return { byLinkId, parentPath };
}

export function inferSourceGroupLinkId(questionnaire: Questionnaire | undefined | null, columns: Column[]): string | undefined {
  if (!questionnaire || columns.length === 0) return undefined;

  const { parentPath } = indexQuestionnaire(questionnaire);

  const first = columns[0].sourceLinkId;
  const firstAncestors = parentPath.get(first);
  if (!firstAncestors || firstAncestors.length === 0) return undefined;

  const otherChains = columns.slice(1).map(c => parentPath.get(c.sourceLinkId) ?? []);
  const includesInAll = (node: QuestionnaireItem): boolean => otherChains.every(chain => chain.includes(node));

  // 1) deepest common ancestor **repeating** group
  for (let i = firstAncestors.length - 1; i >= 0; i--) {
    const node = firstAncestors[i];
    if (node.type === 'group' && node.repeats && includesInAll(node)) {
      return node.linkId;
    }
  }
  // 2) deepest common ancestor group (non-repeating fallback)
  for (let i = firstAncestors.length - 1; i >= 0; i--) {
    const node = firstAncestors[i];
    if (node.type === 'group' && includesInAll(node)) {
      return node.linkId;
    }
  }
  return undefined;
}

/* -------------------------------------------------------------------------- */
/* QR TRAVERSAL + ROW BUILDING                                               */
/* -------------------------------------------------------------------------- */

export function collectAnswersWithin(root: QuestionnaireResponseItem, targetLinkId: string): string[] {
  const out: string[] = [];
  (function dfs(items?: QuestionnaireResponseItem[]): void {
    if (!items) return;
    for (const item of items) {
      if (item.linkId === targetLinkId && item.answer?.length) {
        for (const a of item.answer) out.push(...answerToStrings(a));
      }
      if (item.item?.length) dfs(item.item);
      if (item.answer?.length) for (const a of item.answer) dfs(a.item);
    }
  })(root.item);
  return out;
}

/**
 * Build rows from QuestionnaireResponse
 * using the inferred source group and column definitions.
 * Keys are stored under `summaryLinkId`
 */
export function buildGTableRows(
  qr: QuestionnaireResponse | undefined | null,
  sourceGroupLinkId: string | undefined,
  columns: Column[],
  opts: { joinMultiple?: (vals: string[]) => string } = {}
): Array<Record<string, string | undefined>> {
  const join = opts.joinMultiple ?? ((vals: string[]): string => vals.join(', '));
  const groupOccurrences = getResponseItemsWithLinkId(sourceGroupLinkId, qr?.item);

  return groupOccurrences.map((occ, idx) => {
    const row: Record<string, string | undefined> = { _rowIndex: String(idx + 1) };
    for (const col of columns) {
      const vals = collectAnswersWithin(occ, col.sourceLinkId);
      row[col.summaryLinkId] = vals.length ? join(vals) : undefined;
    }
    return row;
  });
}

/* -------------------------------------------------------------------------- */
/* ANSWER VALUE EXTRACTION                                                    */
/* -------------------------------------------------------------------------- */

export function extractValueFromCoding(coding: Coding | undefined, field: CodingKeys = 'display'): string {
  if (!coding) return '';
  return coding[field] ?? '';
}

export const extractValueFromQuantity = (quantity: Quantity | undefined, field?: QuantityKeys): string | number => {
  if (!quantity) return '';
  if (field) {
    switch (field) {
      case 'display':
        return `${quantity.value ?? 0} ${quantity.unit ?? ''}`.replace(/'/g, '');
      case 'value':
        return quantity.value ?? 0;
      case 'unit':
        return quantity.unit ?? '';
      case 'system':
        return quantity.system ?? '';
      case 'code':
        return quantity.code ?? '';
      default:
        return '';
    }
  } else {
    return `${quantity.value ?? 0} ${quantity.unit ?? ''}`.replace(/'/g, '');
  }
};

export const extractValueFromDate = (inputValue?: string): string => {
  try {
    if (!inputValue && !isValid(inputValue)) return '';
    return inputValue ? format(inputValue, DATEFORMATS.DATE) : '';
  } catch {
    return '';
  }
};

export const extractValueFromTime = (inputTime?: string): string => {
  if (!inputTime) return '';
  const time = inputTime.split(TIME_SEPARATOR);
  if (time.length !== 3) return '';
  return `${time[0]}${TIME_SEPARATOR}${time[1]}`.trim();
};

export const extractValueFromDateTime = (inputValue?: string): string => {
  try {
    if (!inputValue) return '';
    return inputValue ? format(inputValue, DATEFORMATS.DATETIME) : '';
  } catch {
    return '';
  }
};

export function answerToStrings(a: QuestionnaireResponseItemAnswer): string[] {
  if (a.valueString != null) return [a.valueString];
  if (a.valueBoolean != null) return [a.valueBoolean === true ? '[X]' : '[ ]'];
  if (a.valueInteger != null) return [String(a.valueInteger)];
  if (a.valueDecimal != null) return [String(a.valueDecimal)];
  if (a.valueDate != null) return [extractValueFromDate(a.valueDate)];
  if (a.valueDateTime != null) return [extractValueFromDateTime(a.valueDateTime)];
  if (a.valueTime != null) return [extractValueFromTime(a.valueTime)];
  if (a.valueCoding) return [extractValueFromCoding(a.valueCoding, 'display')];
  if (a.valueQuantity) return [String(extractValueFromQuantity(a.valueQuantity))];
  return [];
}
