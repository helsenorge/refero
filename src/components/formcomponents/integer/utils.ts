import { QuestionnaireItem } from 'fhir/r4';
import { UseFormRegisterReturn } from 'react-hook-form';

import { isRequired } from '../../../util';
import { getMaxValueExtensionValue, getMinValueExtensionValue, getValidationTextExtension } from '../../../util/extension';
import { Resources } from '../../../util/resources';
import { FormProps } from '../../../validation/ReactHookFormHoc';

export const integerFormRegister = (
  register: FormProps['register'],
  item: QuestionnaireItem,
  resources?: Resources,
  value?: string | number | number[] | undefined,
  onChange?: (event: React.FormEvent<HTMLInputElement>) => void
): UseFormRegisterReturn => {
  return register(item.linkId, {
    required: {
      value: isRequired(item),
      message: resources?.formRequiredErrorMessage ?? 'Feltet er påkrevd',
    },
    max: {
      value: Number(getMaxValueExtensionValue(item)),
      message: getValidationTextExtension(item) ?? resources?.oppgiGyldigVerdi ?? 'Verdien er for høy',
    },
    min: {
      value: Number(getMinValueExtensionValue(item)),
      message: getValidationTextExtension(item) ?? resources?.oppgiGyldigVerdi ?? 'Verdien er for lav',
    },
    value,
    onChange,
  });
};
