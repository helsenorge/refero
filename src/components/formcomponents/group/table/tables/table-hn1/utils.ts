import { QuestionnaireItem, QuestionnaireResponse, QuestionnaireResponseItem } from 'fhir/r4';

import { ITableH1, ITableH1Column, ITableH1Row, TABLE_HN1_TABLE_TYPES } from './interface';
import ItemType from '@/constants/itemType';
import { QuestionnaireItemWithAnswers } from '../interface';
import { getEnabledQuestionnaireItemsWithAnswers, transformAnswersToListOfStrings } from '../utils';

export const getTableHN1bodyObject = (
  items: QuestionnaireItemWithAnswers[],
  questionnaireResponse?: QuestionnaireResponse | null
): ITableH1 => {
  if (!questionnaireResponse || items.length === 0) {
    return [];
  }
  const answerItems = getEnabledQuestionnaireItemsWithAnswers(items, questionnaireResponse);
  const processItems = (items: QuestionnaireResponseItem[]): ITableH1Row[] => {
    return items.reduce((acc: ITableH1Row[], item) => {
      const row = transformItemToHN1Row(item);
      const newAcc = acc.concat(row);

      const nestedRows = item.item ? processItems(item.item) : [];
      return newAcc.concat(nestedRows);
    }, []);
  };

  return processItems(answerItems);
};

export const createColumnsFromAnswers = (item: QuestionnaireResponseItem): ITableH1Column[] => {
  const combinedAnswer = transformAnswersToListOfStrings((item as QuestionnaireItem).type, item.answer || []).join(', ');
  const columns: ITableH1Column[] = [
    {
      id: `${item.linkId}-question`,
      type: (item as QuestionnaireItem).type === ItemType.GROUP ? TABLE_HN1_TABLE_TYPES.HEADER : TABLE_HN1_TABLE_TYPES.VALUE,
      value: item.text || '',
    },
    {
      value: combinedAnswer,
      type: TABLE_HN1_TABLE_TYPES.VALUE,
      id: `${item.linkId}-answer`,
    },
  ];

  return columns;
};
export const transformItemToHN1Row = (item: QuestionnaireResponseItem): ITableH1Row => {
  const columns = createColumnsFromAnswers(item);

  return {
    id: item.linkId,
    columns: columns,
  };
};
