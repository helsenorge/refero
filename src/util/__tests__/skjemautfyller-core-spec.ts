import * as chai from 'chai';
import {
  getRootQuestionnaireResponseItemFromData,
  getQuestionnaireResponseItemWithLinkid,
  getAnswerFromResponseItem,
  hasAnswer,
  getResponseItemWithPath,
  getItemWithTypeFromArray,
  enableWhenMatchesAnswer,
  createPathForItem,
  Path,
} from '../refero-core';
import {
  QuestionnaireResponseItem,
  QuestionnaireResponseItemAnswer,
  QuestionnaireItemEnableWhen,
  QuestionnaireItem,
} from '../../types/fhir';
import itemType from '../../constants/itemType';
import { pathify } from '../../reducers/__tests__/utils';
import { dataModel } from './__data__/testDataModel';

const should = chai.should();

describe('getQuestionnaireResponseItemFromData', () => {
  it('should not fail on empty form', () => {
    const response = getRootQuestionnaireResponseItemFromData('123123', {
      Content: null,
    });
    expect(response).toEqual(undefined);
  });
  it('should return response item', () => {
    const item = getRootQuestionnaireResponseItemFromData('decimal2', dataModel.refero.form.FormData);
    should.exist(item);
  });
});

describe('get questionnaireResponse item', () => {
  const data = {
    linkId: '123',
    item: [
      {
        linkId: '456',
        item: [
          {
            linkId: 'testId',
          },
        ],
      },
    ],
    answer: [
      {
        item: [
          {
            linkId: '789',
            item: [
              {
                linkId: 'test',
              },
            ],
          },
        ],
      },
    ],
  } as QuestionnaireResponseItem;

  it('should return correct item', () => {
    //should.not.exist(getQuestionnaireResponseItemWithLinkid('123', undefined, false, []));
    should.not.exist(getQuestionnaireResponseItemWithLinkid('123', {} as QuestionnaireResponseItem, []));
    should.exist(getQuestionnaireResponseItemWithLinkid('123', data, []));
    should.not.exist(getQuestionnaireResponseItemWithLinkid('753159', data, []));
    let item = getQuestionnaireResponseItemWithLinkid('456', data, pathify('123', '456'));
    should.exist(item);
    if (item) {
      item.linkId.should.equal('456');
    }
    item = getQuestionnaireResponseItemWithLinkid('789', data, pathify('123', '789'));
    should.exist(item);
    if (item) {
      item.linkId.should.equal('789');
    }
    item = getQuestionnaireResponseItemWithLinkid('test', data, pathify('123', '789', 'test'));
    should.exist(item);
    if (item) {
      item.linkId.should.equal('test');
    }
    item = getQuestionnaireResponseItemWithLinkid('testId', data, pathify('123', '456', 'testId'));
    should.exist(item);
    if (item) {
      item.linkId.should.equal('testId');
    }
  });

  it('should return correct item for repeating elements', () => {
    const dataCopy = JSON.parse(JSON.stringify(data)) as QuestionnaireResponseItem;
    if (dataCopy.item && dataCopy.item[0].item) {
      dataCopy.item[0].item[0].answer = [{ valueString: '2' } as QuestionnaireResponseItemAnswer];
    }
    const repeatingData = { item: [data, dataCopy], linkId: '1' } as QuestionnaireResponseItem;

    let item = getQuestionnaireResponseItemWithLinkid('testId', repeatingData, [
      { linkId: '1' },
      { linkId: '123', index: 1 },
      { linkId: '456' },
      { linkId: 'testId' },
    ]);
    should.exist(item);
    if (item) {
      item.linkId.should.equal('testId');
      should.exist(item.answer);
      if (item.answer && item.answer[0].valueString) {
        item.answer[0].valueString.should.equal('2');
      }
    }
  });
});

describe('get answer from questionnareReponse', () => {
  const data: QuestionnaireResponseItem = {
    linkId: '123',
    answer: [
      {
        valueString: 'abc',
        item: [
          {
            linkId: '789',
          },
        ],
      },
      {
        valueString: 'def',
      },
    ],
  } as QuestionnaireResponseItem;

  it('should return correct answer', () => {
    should.not.exist(getAnswerFromResponseItem(undefined));
    should.not.exist(getAnswerFromResponseItem({} as QuestionnaireResponseItem));
    const answer = getAnswerFromResponseItem(data);
    should.exist(answer);
    if (answer && 'valueString' in answer && answer.valueString) {
      answer.valueString.should.equal('abc');
    }
  });
});

describe('hasAnswer', () => {
  it('should return no answer on invalid input', () => {
    expect(hasAnswer({} as QuestionnaireResponseItemAnswer)).toEqual(false);
    expect(hasAnswer({ valueCoding: {} } as QuestionnaireResponseItemAnswer)).toEqual(false);
    expect(hasAnswer({ valueDate: undefined } as QuestionnaireResponseItemAnswer)).toEqual(false);
    expect(hasAnswer({ valueDateTime: undefined } as QuestionnaireResponseItemAnswer)).toEqual(false);
    expect(hasAnswer({ valueDecimal: undefined } as QuestionnaireResponseItemAnswer)).toEqual(false);
    expect(hasAnswer({ valueInteger: undefined } as QuestionnaireResponseItemAnswer)).toEqual(false);
    expect(hasAnswer({ valueString: '' } as QuestionnaireResponseItemAnswer)).toEqual(false);
    expect(hasAnswer({ valueString: undefined } as QuestionnaireResponseItemAnswer)).toEqual(false);
    expect(hasAnswer({ valueTime: undefined } as QuestionnaireResponseItemAnswer)).toEqual(false);
  });

  it('should return true if any answer type is given', () => {
    expect(hasAnswer({ valueBoolean: true } as QuestionnaireResponseItemAnswer)).toEqual(true);
    expect(hasAnswer({ valueBoolean: false } as QuestionnaireResponseItemAnswer)).toEqual(true);
    expect(hasAnswer({ valueString: 'asdf' } as QuestionnaireResponseItemAnswer)).toEqual(true);
  });
});

describe('item path', () => {
  it('should create path correct', () => {
    expect(
      createPathForItem([], { linkId: 'qwe', repeats: true } as QuestionnaireItem, { linkId: 'qwe' } as QuestionnaireResponseItem, 1)
    ).toEqual([{ linkId: 'qwe', index: 1 }]);
    expect(
      createPathForItem([], { linkId: 'qwe' } as QuestionnaireItem, { linkId: 'qwe' } as QuestionnaireResponseItem, undefined)
    ).toEqual([{ linkId: 'qwe' }]);
    let response: Array<Path> = [
      { linkId: 'test', index: 0 },
      { linkId: 'abc', index: 2 },
    ];
    expect(
      createPathForItem(
        [{ linkId: 'test', index: 0 }],
        { linkId: 'abc', repeats: true } as QuestionnaireItem,
        { linkId: 'abc' } as QuestionnaireResponseItem,
        2
      )
    ).toEqual(response);
    const path = [{ linkId: 'test' }, { linkId: 'test2' }];
    response = [{ linkId: 'test' }, { linkId: 'test2' }, { linkId: 'abc', index: 3 }];
    expect(
      createPathForItem(path, { linkId: 'abc', repeats: true } as QuestionnaireItem, { linkId: 'abc' } as QuestionnaireResponseItem, 3)
    ).toEqual(response);
  });
});

describe('get item with path', () => {
  it('should return undefined on invalid data', () => {
    expect(getResponseItemWithPath([], dataModel.refero.form.FormData)).toBeUndefined();
  });

  it('should return undefined on non existent path', () => {
    expect(getResponseItemWithPath([{ linkId: 'markus' }], dataModel.refero.form.FormData)).toBeUndefined();
  });

  it('should return data', () => {
    let item = getResponseItemWithPath([{ linkId: 't0' }], dataModel.refero.form.FormData);
    should.exist(item);
    if (!item) {
      fail();
      return;
    }
    expect(item.linkId).toEqual('t0');

    let path = [
      {
        linkId: 't0',
      },
      {
        linkId: 't11',
      },
      {
        linkId: 't21',
      },
    ];
    item = getResponseItemWithPath(path, dataModel.refero.form.FormData);
    should.exist(item);
    if (!item) {
      fail();
      return;
    }
    expect(item.linkId).toEqual('t21');

    path = [
      {
        linkId: 't0',
      },
      {
        linkId: 't11',
      },
      {
        linkId: 't21',
      },
      {
        linkId: 't31.abc',
      },
    ];
    item = getResponseItemWithPath(path, dataModel.refero.form.FormData);
    should.exist(item);
    if (!item) {
      fail();
      return;
    }
    expect(item.linkId).toEqual('t31.abc');

    path = [
      {
        linkId: 't0',
      },
      {
        linkId: 't11',
      },
      {
        linkId: 't21',
      },
      {
        linkId: 't32.def',
      },
      {
        linkId: 't41',
      },
    ];
    item = getResponseItemWithPath(path, dataModel.refero.form.FormData);
    should.exist(item);
    if (!item) {
      fail();
      return;
    }
    expect(item.linkId).toEqual('t41');

    path = [
      {
        linkId: 't0',
      },
    ];
    item = getResponseItemWithPath(path, dataModel.refero.form.FormData);
    should.exist(item);
    if (!item) {
      fail();
      return;
    }
    expect(item.linkId).toEqual('t0');

    path = [
      {
        linkId: 'group1',
      },
      {
        linkId: 'group1.1',
      },
    ];
    item = getResponseItemWithPath(path, dataModel.refero.form.FormData);
    should.exist(item);
    if (!item) {
      fail();
      return;
    }
    expect(item.linkId).toEqual('group1.1');
  });
});

describe('getItemWithTypeFromArray', () => {
  it('should return correct items', () => {
    const item1: QuestionnaireResponseItem = {
      linkId: '123',
      answer: [
        {
          valueAttachment: { title: '' },
        },
      ],
    } as QuestionnaireResponseItem;
    const item2: QuestionnaireResponseItem = {
      linkId: '123',
      answer: [
        {
          valueString: 'lala',
        },
      ],
    } as QuestionnaireResponseItem;
    const response = getItemWithTypeFromArray(itemType.ATTATCHMENT, [item1, item2]);
    expect(response).toMatchSnapshot();
  });

  it('should not fail on empty answer array', () => {
    const item1: QuestionnaireResponseItem = {
      linkId: '123',
      answer: [],
    } as QuestionnaireResponseItem;
    const response = getItemWithTypeFromArray(itemType.ATTATCHMENT, [item1]);
    expect(response).toMatchSnapshot();
  });
});
describe('Given a Questionnaire with operator="exists" and answerBoolean=false', () => {
  it('When an item does not have an answer then enablewhen should return true', () => {
    const result = enableWhenMatchesAnswer(
      { question: '1.0.0', operator: 'exists', answerBoolean: false } as QuestionnaireItemEnableWhen,
      undefined
    );
    expect(result).toBe(true);
  });

  it('When an item does have an answer then enablewhen should return false', () => {
    const result = enableWhenMatchesAnswer({ question: '1.0.0', operator: 'exists', answerBoolean: false } as QuestionnaireItemEnableWhen, [
      { valueString: 'et svar' } as QuestionnaireResponseItemAnswer,
    ]);
    expect(result).toBe(false);
  });
});
