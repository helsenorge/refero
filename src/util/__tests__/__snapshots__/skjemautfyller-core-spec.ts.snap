// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getItemWithTypeFromArray should not fail on empty answer array 1`] = `undefined`;

exports[`getItemWithTypeFromArray should return correct items 1`] = `
Array [
  Object {
    "answer": Array [
      Object {
        "valueAttachment": Object {
          "title": "",
        },
      },
    ],
    "linkId": "123",
  },
]
`;
